module/zfs/zfs_ctldir.c

```diff
@@ -566,10 +566,14 @@ zfsctl_destroy(zfsvfs_t *zfsvfs)
 		spa_t *spa = zfsvfs->z_os->os_spa;
 		uint64_t objsetid = dmu_objset_id(zfsvfs->z_os);
 
-		rw_enter(&zfs_snapshot_lock, RW_WRITER);
+		rw_enter(&zfs_snapshot_lock, RW_READER);
 		if ((se = zfsctl_snapshot_find_by_objsetid(spa, objsetid))
 		    != NULL) {
 			zfsctl_snapshot_unmount_cancel(se);
+			if (!rw_tryupgrade(&zfs_snapshot_lock)) {
+				rw_exit(&zfs_snapshot_lock);
```

Comment:
My concern with this is that dropping and reacquiring the lock here it leaves a small window where the `zfs_snapentry_t` can still be found on the `_by_name` and `_by_objsetid` list just before it's going to be destroyed. 

Since the `RW_WRITER` lock is there to protect the lookup and and list removal. it makes sense to do both of those things first while holding the lock.  Then drop the lock, and call `zfsctl_snapshot_unmount_cancel`.  Since the validity of the `se` is being protected by the `se->se_refcount`.  It looks like you will need to remove the ASSERT from `zfsctl_snapshot_unmount_cancel`.

Reply:
Fair concern, reworked the change as suggested.

The test case is to reproduce actually, needs to add a sleep for few seconds in snapentry_expire(), before se->se_taskqid = TASKQID_INVALID; and during this sleep the zfs umount zfs_name( for which the snapshot existed/created) has to be fired.
But this sleep is only needed for testing purposes. But i do not see any flags under which I can add this sleep. Any idea?